src/storage/v2/indices/text_index.cpp

```diff
@@ -0,0 +1,315 @@
+    }
+  }
+
+  return serialized_properties;
+}
+
+std::vector<mgcxx::text_search::Context *> TextIndex::GetApplicableTextIndices(const std::vector<LabelId> &labels) {
+  if (!flags::AreExperimentsEnabled(flags::Experiments::TEXT_SEARCH)) {
+    throw query::TextSearchDisabledException();
+  }
+
+  std::vector<mgcxx::text_search::Context *> applicable_text_indices;
+  for (const auto &label : labels) {
+    if (label_to_index_.contains(label)) {
+      applicable_text_indices.push_back(&index_.at(label_to_index_.at(label)).context_);
+    }
+  }
+  return applicable_text_indices;
+}
+
+std::vector<mgcxx::text_search::Context *> TextIndex::GetApplicableTextIndices(Vertex *vertex) {
+  return GetApplicableTextIndices(vertex->labels);
+}
+
+void TextIndex::LoadNodeToTextIndices(const std::int64_t gid, const nlohmann::json &properties,
+                                      const std::vector<mgcxx::text_search::Context *> &applicable_text_indices) {
+  // NOTE: Text indexes are presently all-property indices. If we allow text indexes restricted to specific properties,
+  // an indexable document should be created for each applicable index.
+  nlohmann::json document = {};
+  document["data"] = properties;
+  document["metadata"] = {};
+  document["metadata"]["gid"] = gid;
+  document["metadata"]["deleted"] = false;
+  document["metadata"]["is_node"] = true;
+
+  for (auto *index_context : applicable_text_indices) {
+    try {
+      mgcxx::text_search::add_document(
+          *index_context,
+          mgcxx::text_search::DocumentInput{
+              .data = document.dump(-1, ' ', false, nlohmann::json::error_handler_t::replace)},
+          kDoSkipCommit);
+    } catch (const std::exception &e) {
+      throw query::TextSearchException("Tantivy error: {}", e.what());
+    }
+  }
+}
+
+void TextIndex::CommitLoadedNodes(mgcxx::text_search::Context &index_context) {
+  // As CREATE TEXT INDEX (...) queries don’t accumulate deltas, db_transactional_accessor_->Commit() does not reach
+  // the code area where changes to indices are committed. To get around that without needing to commit text indices
+  // after every such query, we commit here.
+  try {
+    mgcxx::text_search::commit(index_context);
+  } catch (const std::exception &e) {
+    throw query::TextSearchException("Tantivy error: {}", e.what());
+  }
+}
+
+void TextIndex::AddNode(Vertex *vertex_after_update, NameIdMapper *name_id_mapper,
+                        const std::vector<mgcxx::text_search::Context *> &applicable_text_indices) {
+  if (!flags::AreExperimentsEnabled(flags::Experiments::TEXT_SEARCH)) {
+    throw query::TextSearchDisabledException();
+  }
+
+  auto vertex_properties = vertex_after_update->properties.Properties();
+  LoadNodeToTextIndices(vertex_after_update->gid.AsInt(), SerializeProperties(vertex_properties, name_id_mapper),
+                        applicable_text_indices);
+}
+
+void TextIndex::AddNode(Vertex *vertex_after_update, NameIdMapper *name_id_mapper) {
```

Comment:
No need for both of these methods. Either always pass applicable_text_indices or never. If you can always get applicable_text_indices by calling GetApplicableTextIndices then do that in the method and solved

Reply:
> Your other suggestion (always get applicable_text_indices by calling GetApplicableTextIndices) works, but it’s suboptimal for updating nodes because we'd have to call GetApplicableTextIndices redundantly: once when calling RemoveNode, and once when calling AddNode.

Then method can have optional arg text_indices -> if you pass something, use it, if not do it by your own in the method.

Reply:
While this is a good idea, this way we can’t distinguish whether the `applicable_text_indices` argument is empty because a) no text indices are applicable b) a value wasn’t provided.

Reply:
optional is for that. Using empty vector or empty string for such purposes isn't best

Reply:
Done